### PR TITLE
Add arrow navigation component and set visibility of List component scrollbar

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1280) |

Associated PR: https://github.com/pi-top/pi-topOS-Web-Portal/pull/377

#### Main changes

- Add `ArrowNavigationIndicator` component.
- Add `visible_scrollbar` property to `List` component to set the visibility of the components scrollbar.
- Allow `List` to navigate more than 1 page at the time using the `distance` parameter.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: pt_miniscreen/components/arrow_navigation_indicator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class ArrowPosition(Enum):
+    TOP = auto()
+    BOTTOM = auto()
+
+
+class ArrowDrawing(Component):
```
Comment (olivierwilkinson) [inline pt_miniscreen/components/arrow_navigation_indicator.py]: I'm not sure this needs to be a component, since the state is only in charge of position and whether to render or not which the parent can control already. Instead a "render function" similar to `rectangle` might make more sense:

```python3
def arrow(image, rotation=0, fill=1):
    center = (int(image.width / 2), int(image.height / 2))
    radius = int(min(image.width, image.height) / 2)
    ImageDraw.Draw(image).regular_polygon((center, radius), 3, fill=1, rotation=rotation)
    return image
```

then it can be used by `ArrowNavigationIndicator` like:
```python3
class ArrowNavigationIndicator(Component):
    def __init__(
        self,
        upper_arrow_padding=(0, 0),
        lower_arrow_padding=(0, 0),
        top_arrow_visible=True,
        bottom_arrow_visible=True,
        **kwargs
    ):
        super().__init__(
            **kwargs,
            initial_state={
                "upper_arrow_padding": upper_arrow_padding,
                "lower_arrow_padding": lower_arrow_padding,
                "top_arrow_visible": top_arrow_visible,
                "bottom_arrow_visible": bottom_arrow_visible
            },
        )

    def render(self, image):
        layers = []
        arrow_size = (10, 10)

        if self.state["upper_arrow_visible"]:
            layers.append(
                layer(
                    arrow,
                    size=arrow_size,
                    pos=self.state["upper_arrow_padding"],
                )
            )

        if self.state["bottom_arrow_visible"]:
            icon_padding = self.state["lower_arrow_padding"]
            icon_pos = (icon_padding[0], image.height - arrow_size[1] - icon_padding[1])

            layers.append(
                layer(partial(arrow, rotation=180), size=arrow_size, pos=icon_pos)
            )

        return apply_layers(image, layers)
```

I suspect there was also some 'hidden' padding because the arrow component size and the actual arrow size didn't match up. So yeah "arrow_size" might need to change for this to be the same as before.

Comment (olivierwilkinson) [reply pt_miniscreen/components/arrow_navigation_indicator.py]: Side note: I wasn't really sure what to call these "render functions". I feel they need a name of some sort so that if they were in their own file they might be used like:

```python3
from pt_miniscreen.core.utils import render
...
class Test(Component):
    def render(self, image):
        return apply_layers(image, [
            layer(render.arrow), # <- or like render.rectangle or something.
        ])
```

maybe "paint" or "draw" is better? Not sure, would be interested in your opinion here!

```python3
from pt_miniscreen.core.utils import paint
...
class Test(Component):
    def render(self, image):
        return apply_layers(image, [
            layer(paint.arrow),
        ])
```

File: pt_miniscreen/core/components/list.py [hunk 2]
```diff
@@ -53,12 +53,31 @@ def __init__(
 
         self._rows_snapshot = None
         self._cleanup_transition = threading.Event()
+        self.visible_scrollbar = True
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:56]: I think this should be in state, no reason for someone not to be able to toggle it and it directly relates to the rendered output.

File: pt_miniscreen/core/components/stack.py [hunk 3]
```diff
@@ -52,6 +52,10 @@ def active_index(self):
         except ValueError:
             return None
 
+    @property
+    def components(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/stack.py:56]: I think this should be called `stack`, `components` is quite vague.

File: pt_miniscreen/core/components/list.py [hunk 4]
```diff
@@ -91,43 +110,40 @@ def _scroll_transition(self):
         self._rows_snapshot = None
         self.state.update({"active_transition": None, "transition_progress": 0})
 
-    def scroll_up(self):
+    def scroll_to(self, direction, distance=1):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:110]: I'm not sure about this distance behaviour. It seems that if you scroll with a distance greater than 1 the order or self.rows no longer matches up with the Rows array and so it will behave pretty wierdly, particularly in the case where there are more than one visible row!

To scroll more the 1 row we should instantiate and add all the rows that need to be scrolled through and then remove the ones that are no longer visible when the transition ends.

File: pt_miniscreen/core/components/list.py [hunk 5]
```diff
@@ -91,43 +110,40 @@ def _scroll_transition(self):
...
-        Rows = self.state["Rows"]
-        next_top_row_index = self.state["top_row_index"] + 1
-        max_top_row_index = len(Rows) - self.state["num_visible_rows"]
-        if next_top_row_index > max_top_row_index:
-            logger.debug(f"{self} has no more rows, ignoring scroll down")
-            return
-
-        Row = Rows[next_top_row_index + self.state["num_visible_rows"] - 1]
-        self.rows.append(self.create_child(Row))
+    def scroll_up(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:139]: I think this can have distance in as an argument as well, seems like a good API to have!

File: pt_miniscreen/core/components/list.py [hunk 6]
```diff
@@ -60,6 +60,11 @@ def __init__(
         end_index = self.state["num_visible_rows"] + self.state["top_row_index"]
         self.rows = [self.create_child(Row) for Row in Rows[start_index:end_index]]
 
+    @property
+    def components_to_top_row(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:64]: This is used to figure out the scroll distance needed to scroll to the bottom right? Maybe something like "distance_to_bottom" and "distance_to_top" would be a bit friendlier. That terminology could also be used in "scroll_to_bottom" and "scroll_to_top" methods.

**Review by olivierwilkinson — CHANGES_REQUESTED** [review]
There are also some other areas that could be tweaked but overall I like the direction :D. I've requested changes because of the scrolling behaviour when going a distance greater than one. 

there are some areas I might tweak and the implementation of scrolling more than one row at a time needs to take into account

File: pt_miniscreen/core/components/list.py [hunk 7]
```diff
@@ -53,12 +53,31 @@ def __init__(
...
+    @property
+    def visible_scrollbar(self):
+        return self._visible
+
+    @visible_scrollbar.setter
+    def visible_scrollbar(self, value):
+        self._visible = value
+
+    @property
+    def can_scroll_down(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:69]: I like these as properties by the way, seems like a really common thing to want to know!

File: pt_miniscreen/core/components/list.py (hunk 2 above)
Comment (jcapona) [reply pt_miniscreen/core/components/list.py:56]: good point!

File: pt_miniscreen/components/arrow_navigation_indicator.py (hunk 1 above)
Comment (jcapona) [reply pt_miniscreen/components/arrow_navigation_indicator.py]: `draw` is a good name ... I think using these functions as `draw.rectangle` or `draw.arrow` makes sense.... I also thought in `shapes` but `draw` is more comprehensive and would let us add more 'render functions' into that file later on

File: pt_miniscreen/core/components/list.py [hunk 8]
```diff
@@ -59,6 +61,33 @@ def __init__(
...
+        next_top_row_index = self.state["top_row_index"] + 1
+        max_top_row_index = len(self.state["Rows"]) - self.state["num_visible_rows"]
+        return next_top_row_index <= max_top_row_index
+
+    @property
+    def can_scroll_up(self):
+        return self.state["top_row_index"] > 0
+
     @property
     def visible_rows(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:92]: I think visible_rows needs to be updated to account for scroll distance, right now it assumes a scroll distance of 1

File: pt_miniscreen/core/components/list.py [hunk 9]
```diff
@@ -59,6 +61,33 @@ def __init__(
...
+    def distance_to_bottom(self):
+        max_top_row_index = len(self.state["Rows"]) - self.state["num_visible_rows"]
+        return max_top_row_index - self.state["top_row_index"]
+
+    @property
+    def distance_to_top(self):
+        return self.state["top_row_index"]
+
+    @property
+    def can_scroll_down(self):
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:82]: These properties assume a scroll distance of 1. It's correct to say if we can scroll down once then we can scroll down but we should probably change them to be functions that also take a distance argument. Later we use these to see if we can scroll in the `scroll_to` function but it doesn't give us enough info (I'll add a comment there with some more detail.)

File: pt_miniscreen/core/components/list.py [hunk 10]
```diff
@@ -85,49 +114,51 @@ def _scroll_transition(self):
...
-        next_top_row_index = self.state["top_row_index"] - 1
-        if next_top_row_index < 0:
-            logger.debug(f"{self} has no more rows, ignoring scroll up")
+            logger.info(f"{self} currently scrolling, ignoring scroll")
             return
 
-        Row = self.state["Rows"][next_top_row_index]
-        self.rows.insert(0, self.create_child(Row))
+        if direction == "UP":
+            if not self.can_scroll_up:
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:134]: When scrolling with a distance greater than one it's possible to get an out of range exception since this assumes the scroll distance will be one. If it took the distance as an argument this would work well :)

```
        if direction == "UP":
            if not self.can_scroll_up(distance):
```

File: pt_miniscreen/core/components/list.py [hunk 11]
```diff
@@ -85,49 +114,51 @@ def _scroll_transition(self):
...
-        self.rows.remove(row_to_remove)
-        self.remove_child(row_to_remove)
+        if active_transition == "DOWN":
+            rows_to_remove = self.rows[:distance]
+        else:
+            rows_to_remove = self.rows[-distance:]
+
+        for row in rows_to_remove:
+            self.rows.remove(row)
+            self.remove_child(row)
```
Comment (olivierwilkinson) [inline pt_miniscreen/core/components/list.py:124]: If visible_rows reflected the final rows array (see my other comment) then this could leverage it to be a little more concise and descriptive.

```suggestion
        for row in filter(lambda row: row not in self.visible_rows, self.rows):
            self.rows.remove(row)
            self.remove_child(row)
```

**Review by olivierwilkinson — COMMENTED** [review]
This is much closer! When trying it out I noticed a couple things though:
- scrolling down animates well regardless of distance which was great to see
- scrolling up has an off by one issue with the animation when scrolling a distance greater than one, the actual adding and removing logic is sound though.
- the scrollbar animation is also a bit off with distances greater than one.

I also noticed a couple more things when investigating the code which I've added comments on. I didn't dig into the animation issues but would be up for pairing on it if you like 😄 

https://user-images.githubusercontent.com/15321261/165070444-d2069c5a-0a13-4922-8185-0da097c81c36.mov

**Comment by olivierwilkinson** [comment]
Here is a snippet for the list I tested with in the video:
```python3
class TestRoot(List):
    def __init__(self, **kwargs):
        super().__init__(**kwargs, num_visible_rows=4, Rows=[
            partial(Text, text=f"{index + 1}", align="center", vertical_align="center") for index in range(0, 30)
        ])
```

**Comment by olivierwilkinson** [comment]
Also, just something to chew on is that the ArrowNavigationIndicator component is quite specific, it might be better to incorperate it into the List component in some way since they are closely related. Maybe a "navigation_indicator" arg with "SCROLLBAR" and "ARROWS" options? What do you think?

File: pt_miniscreen/components/arrow_navigation_indicator.py (hunk 1 above)
Comment (olivierwilkinson) [reply pt_miniscreen/components/arrow_navigation_indicator.py]: yeah draw is tempting, my only worry is that it's too close to the terminology used by PIL for something that behaves differently to these functions. Anyhoo let's address it when the number of these utils get's a bit bigger! No need to overcommit to anything yet. :)

**Review by olivierwilkinson — DISMISSED** [review]
LGTM! Scrolling looks sweet and this makes the list much more generally useful!

The only things that might be worth doing if we have time is to do a quick test for the `ArrowNavigationIndicator` to check we can hide and show the arrows and to check `List.visible_rows` is correct while scrolling a distance greater than 1 👍 

Totally up to you though, I'm happy to merge with the current tests 😄
